clients/fides-js/src/lib/gpp/string-to-consent.ts

```diff
@@ -0,0 +1,308 @@
+  preferences.specialFeatures = (tcfeuv2.SpecialFeatureOptins as boolean[])
+    .map((value, index) => (value ? (index + 1).toString() : null))
+    .filter((value): value is string => value !== null);
+
+  // Get special purposes and features from experience config
+  if ("tcf_special_purposes" in experience) {
+    preferences.specialPurposes =
+      experience.tcf_special_purposes?.map((p) => p.id.toString()) || [];
+    preferences.features =
+      experience.tcf_features?.map((f) => f.id.toString()) || [];
+  } else if ("tcf_special_purpose_ids" in experience) {
+    preferences.specialPurposes =
+      experience.tcf_special_purpose_ids?.map((id) => id.toString()) || [];
+    preferences.features =
+      experience.tcf_feature_ids?.map((id) => id.toString()) || [];
+  }
+
+  // Custom purposes consent remain empty as per example
+  preferences.customPurposesConsent = [];
+
+  return preferences;
+};
+
+interface FidesStringToConsentArgs {
+  fidesString: string;
+  cmpApi: CmpApi;
+}
+export const fidesStringToConsent = ({
+  fidesString,
+  cmpApi,
+}: FidesStringToConsentArgs) => {
+  if (!fidesString || !cmpApi) {
+    return;
+  }
+
+  const { options, cookie, geolocation, locale, config } = window.Fides;
+  const experience = window.Fides.experience as
+    | PrivacyExperience
+    | PrivacyExperienceMinimal;
+  if (
+    !experience ||
+    !experience.experience_config ||
+    !experience.privacy_notices
+  ) {
+    return;
+  }
+
+  const isTCF =
+    experience.experience_config.component === ComponentType.TCF_OVERLAY;
+
+  const { gpp: gppString } = decodeFidesString(fidesString);
+  const fidesRegionString = constructFidesRegionString(geolocation);
+  const matchTranslation = experience.experience_config.translations.find((t) =>
+    areLocalesEqual(t.language, locale),
+  );
+
+  if (!matchTranslation) {
+    return;
+  }
+
+  // Set up GPP string
+  cmpApi.setGppString(gppString);
+  cmpApi.setSignalStatus(SignalStatus.READY);
+  fidesDebugger("GPP: updated GPP from fides_string override", fidesString);
+
+  // Map privacy notices with translations
+  const privacyNoticeItems = experience.privacy_notices.map((notice) => ({
+    notice,
+    bestTranslation:
+      notice.translations.find((t) => areLocalesEqual(t.language, locale)) ||
+      null,
+  }));
+
+  let updateCookie: (oldCookie: FidesCookie) => Promise<FidesCookie>;
+  let tcf: TcfSavePreferences | undefined;
+  let consentPreferencesToSave: SaveConsentPreference[];
+
+  if (!isTCF) {
+    // Handle non-TCF case
+    const consent = getConsentFromGppCmpApi({
+      cmpApi,
+      experience: experience as PrivacyExperience,
+    });
+
+    const enabledPrivacyNoticeKeys = Object.entries(consent)
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      .filter(([_, value]) => value)
+      .map(([key]) => key);
+
+    consentPreferencesToSave = createConsentPreferencesToSave(
```

Comment:
type annotation

Reply:
I generally prefer inferred typing as appropriate. If the TS knows what type it is I find it redundant to tell it again.